# Apply

In [1]:
import numpy as np
import pandas as pd

## download the data and read the data into a variable 
-  student-mat.csv

In [2]:
student_df = pd.read_csv("student-mat - student-mat.csv")
student_df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        6   5   6   6  
1      5        3      3     1     1      3        4   5   5   6  
2      4        3      2     2     3      3       10   7   8  10  
3      3        2      2     1     1      5        2  15  14  15  
4      4        3      2     1     2      5        4   6  10  10  

[5 rows x 33 columns]

##  slice the dataframe from 'school' until the 'guardian' column

In [3]:
selected_data = student_df.loc[:, "school":"guardian"]
selected_data.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  
0  course   mother  
1  course   father  
2   other   mother  
3    home   mother  
4    home   father

## Create a lambda function that will capitalize strings.

In [4]:
capitalize = lambda x: x.capitalize()

## Capitalize both Mjob and Fjob

In [5]:
selected_data["Mjob"]= selected_data["Mjob"].apply(capitalize)

fjob = selected_data["Fjob"] = selected_data["Fjob"].apply(capitalize)

selected_data

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   At_home   Teacher   
1       GP   F   17       U     GT3       T     1     1   At_home     Other   
2       GP   F   15       U     LE3       T     1     1   At_home     Other   
3       GP   F   15       U     GT3       T     4     2    Health  Services   
4       GP   F   16       U     GT3       T     3     3     Other     Other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  Services  Services   
391     MS   M   17       U     LE3       T     3     1  Services  Services   
392     MS   M   21       R     GT3       T     1     1     Other     Other   
393     MS   M   18       R     LE3       T     3     2  Services     Other   
394     MS   M   19       U     LE3       T     1     1     Other   At_home   

     reason guardian  
0    course   mother  
1    course   father  
2     other   mother  
3      home   mother  
4      home   father  
..      ...      ...  
390  course    other  
391  course   mother  
392  course    other  
393  course   mother  
394  course   father  

[395 rows x 12 columns]

## Create a function called majority that returns a boolean value to a new column called legal_drinker (Consider majority as older than 17 years old)

In [6]:
def majority():
    selected_data["legal_drinker"] = selected_data["age"]>17
    return selected_data["legal_drinker"]
selected_data[majority()]
selected_data.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  At_home   Teacher   
1     GP   F   17       U     GT3       T     1     1  At_home     Other   
2     GP   F   15       U     LE3       T     1     1  At_home     Other   
3     GP   F   15       U     GT3       T     4     2   Health  Services   
4     GP   F   16       U     GT3       T     3     3    Other     Other   

   reason guardian  legal_drinker  
0  course   mother           True  
1  course   father          False  
2   other   mother          False  
3    home   mother          False  
4    home   father          False

## Multiply every number of the dataset by 10.


In [7]:
student_df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        6   5   6   6  
1      5        3      3     1     1      3        4   5   5   6  
2      4        3      2     2     3      3       10   7   8  10  
3      3        2      2     1     1      5        2  15  14  15  
4      4        3      2     1     2      5        4   6  10  10  

[5 rows x 33 columns]

In [8]:
numeric_columns = student_df.select_dtypes(include = ['float', 'int64']).columns
student_df[numeric_columns]= student_df[numeric_columns]*10
student_df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F  180       U     GT3       A    40    40  at_home   teacher  ...   
1     GP   F  170       U     GT3       T    10    10  at_home     other  ...   
2     GP   F  150       U     LE3       T    10    10  at_home     other  ...   
3     GP   F  150       U     GT3       T    40    20   health  services  ...   
4     GP   F  160       U     GT3       T    30    30    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences   G1   G2   G3  
0     40       30     40    10    10     30       60   50   60   60  
1     50       30     30    10    10     30       40   50   50   60  
2     40       30     20    20    30     30      100   70   80  100  
3     30       20     20    10    10     50       20  150  140  150  
4     40       30     20    10    20     50       40   60  100  100  

[5 rows x 33 columns]

#  United States - Crime Rates - 1960 - 2014

In [9]:
# laod the data ---US_Crime_Rates_1960_2014.csv

In [10]:
crime = pd.read_csv("US_Crime_Rates_1960_2014 - US_Crime_Rates_1960_2014.csv")
crime.head()

Year  Population    Total  Violent  Property  Murder  Forcible_Rape  \
0  1960   179323175  3384200   288460   3095700    9110          17190   
1  1961   182992000  3488000   289390   3198600    8740          17220   
2  1962   185771000  3752200   301510   3450700    8530          17550   
3  1963   188483000  4109500   316970   3792500    8640          17650   
4  1964   191141000  4564600   364220   4200400    9360          21420   

   Robbery  Aggravated_assault  Burglary  Larceny_Theft  Vehicle_Theft  
0   107840              154320    912100        1855400         328200  
1   106670              156760    949600        1913000         336000  
2   110860              164570    994300        2089600         366800  
3   116470              174210   1086400        2297800         408300  
4   130390              203050   1213200        2514400         472800

## look at the data type

In [11]:
crime.dtypes

Year                  int64
Population            int64
Total                 int64
Violent               int64
Property              int64
Murder                int64
Forcible_Rape         int64
Robbery               int64
Aggravated_assault    int64
Burglary              int64
Larceny_Theft         int64
Vehicle_Theft         int64
dtype: object

## change the colum year from int to datetime64

In [12]:
crime["Year"] = pd.to_datetime(crime["Year"], format ="%Y")
crime.head(2)

Year  Population    Total  Violent  Property  Murder  Forcible_Rape  \
0 1960-01-01   179323175  3384200   288460   3095700    9110          17190   
1 1961-01-01   182992000  3488000   289390   3198600    8740          17220   

   Robbery  Aggravated_assault  Burglary  Larceny_Theft  Vehicle_Theft  
0   107840              154320    912100        1855400         328200  
1   106670              156760    949600        1913000         336000

## set the year column as the index

In [13]:
crime.set_index("Year", inplace = True)

In [14]:
crime.head(2)

Population    Total  Violent  Property  Murder  Forcible_Rape  \
Year                                                                        
1960-01-01   179323175  3384200   288460   3095700    9110          17190   
1961-01-01   182992000  3488000   289390   3198600    8740          17220   

            Robbery  Aggravated_assault  Burglary  Larceny_Theft  \
Year                                                               
1960-01-01   107840              154320    912100        1855400   
1961-01-01   106670              156760    949600        1913000   

            Vehicle_Theft  
Year                       
1960-01-01         328200  
1961-01-01         336000

In [15]:
crime.index.year

Int64Index([1960, 1961, 1962, 1963, 1964, 1965, 1966, 1967, 1968, 1969, 1970,
            1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980, 1981,
            1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992,
            1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003,
            2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014],
           dtype='int64', name='Year')

## Group the year by decades and sum the values

In [16]:
group_data = crime.groupby((crime.index.year//10) * 10).sum()
group_data.head(2)

Population      Total  Violent  Property  Murder  Forcible_Rape  \
Year                                                                    
1960  1915053175   49295900  4134930  45160900  106180         236720   
1970  2121193298  100991600  9607930  91383800  192230         554570   

      Robbery  Aggravated_assault  Burglary  Larceny_Theft  Vehicle_Theft  
Year                                                                       
1960  1633510             2158520  13321100       26547700        5292100  
1970  4159020             4702120  28486000       53157800        9739900

In [17]:
crime.resample("10AS").max()

Population     Total  Violent  Property  Murder  Forcible_Rape  \
Year                                                                         
1960-01-01   201385000   7410900   661870   6749000   14760          37170   
1970-01-01   220099000  12249500  1208030  11041500   21460          76390   
1980-01-01   248239000  14251400  1646040  12605400   23040          94500   
1990-01-01   272690813  14872900  1932270  12961100   24700         109060   
2000-01-01   307006550  11878954  1439480  10455277   17030          95235   
2010-01-01   318857056  10363873  1251248   9112625   14866          85593   

            Robbery  Aggravated_assault  Burglary  Larceny_Theft  \
Year                                                               
1960-01-01   298850              311090   1981900        3888600   
1970-01-01   480700              629480   3327700        6601000   
1980-01-01   592910              951710   3795200        7872400   
1990-01-01   687730             1135610   3157200        8142200   
2000-01-01   447403              911706   2228474        7092267   
2010-01-01   369089              781844   2185140        6204601   

            Vehicle_Theft  
Year                       
1960-01-01         878500  
1970-01-01        1112800  
1980-01-01        1564800  
1990-01-01        1661700  
2000-01-01        1261226  
2010-01-01         739565

## What is the most dangerous decade to live in the US?

In [22]:
# According to total population by year
most_dangerous_decade = group_data[group_data.Total == max(group_data.Total)]
most_dangerous_decade

Population      Total   Violent   Property  Murder  Forcible_Rape  \
Year                                                                      
1990  2612825258  136582146  17527048  119053499  211664         998827   

      Robbery  Aggravated_assault  Burglary  Larceny_Theft  Vehicle_Theft  
Year                                                                       
1990  5748930            10568963  26750015       77679366       14624418

In [35]:
group_data.idxmax()

Population            2000
Total                 1990
Violent               1990
Property              1990
Murder                1990
Forcible_Rape         1990
Robbery               1990
Aggravated_assault    1990
Burglary              1980
Larceny_Theft         1990
Vehicle_Theft         1990
dtype: int64

#  Use the u.user file and assign it to a dataframe 

In [2]:
data = pd.read_csv("u.user", delimiter = '|')
data

user_id  age gender     occupation zip_code
0          1   24      M     technician    85711
1          2   53      F          other    94043
2          3   23      M         writer    32067
3          4   24      M     technician    43537
4          5   33      F          other    15213
..       ...  ...    ...            ...      ...
938      939   26      F        student    33319
939      940   32      M  administrator    02215
940      941   20      M        student    97229
941      942   48      F      librarian    78209
942      943   22      M        student    77841

[943 rows x 5 columns]

##  See the first 5 entries

In [3]:
data.head()

user_id  age gender  occupation zip_code
0        1   24      M  technician    85711
1        2   53      F       other    94043
2        3   23      M      writer    32067
3        4   24      M  technician    43537
4        5   33      F       other    15213

##  What is the number of observations in the dataset?

In [4]:
data.shape[0]

943

There is 943 observations in the dataset.

## What is the number of columns in the dataset?

In [5]:
data.columns.nunique()

5

There is 5 columns in the dataset.

## Print the name of all the columns.

In [6]:
data.columns

Index(['user_id', 'age', 'gender', 'occupation', 'zip_code'], dtype='object')

## How is the dataset indexed?

In [7]:
data.index

RangeIndex(start=0, stop=943, step=1)

Dataset indexed by start from 0, stop at 943 by step 1.

## What is datatype of each cloumn?

In [8]:
data.dtypes

user_id        int64
age            int64
gender        object
occupation    object
zip_code      object
dtype: object

user_id and age columns integer and gender, zip_code and occupation are object or categorical. 

##  How many different occupation are in this dataset?

In [9]:
data["occupation"].nunique()

21

## What is the most frequent occupation?

In [11]:
data["occupation"].value_counts().idxmax()

'student'

## Summarize the DataFrame.

In [12]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 943 entries, 0 to 942
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   user_id     943 non-null    int64 
 1   age         943 non-null    int64 
 2   gender      943 non-null    object
 3   occupation  943 non-null    object
 4   zip_code    943 non-null    object
dtypes: int64(2), object(3)
memory usage: 37.0+ KB


## Summarize all the columns

In [13]:
data.describe(include = "all")

user_id         age gender occupation zip_code
count   943.000000  943.000000    943        943      943
unique         NaN         NaN      2         21      795
top            NaN         NaN      M    student    55414
freq           NaN         NaN    670        196        9
mean    472.000000   34.051962    NaN        NaN      NaN
std     272.364951   12.192740    NaN        NaN      NaN
min       1.000000    7.000000    NaN        NaN      NaN
25%     236.500000   25.000000    NaN        NaN      NaN
50%     472.000000   31.000000    NaN        NaN      NaN
75%     707.500000   43.000000    NaN        NaN      NaN
max     943.000000   73.000000    NaN        NaN      NaN

## Summarize only the occupation column

In [14]:
data.occupation.describe()

count         943
unique         21
top       student
freq          196
Name: occupation, dtype: object

## What is the age with least occurrence?

In [15]:
data["age"].value_counts().tail()

7     1
66    1
10    1
11    1
73    1
Name: age, dtype: int64

In [16]:
data["age"].value_counts().idxmin()

7

# Load Json file

## read a raw_nyc_phil.json data

In [18]:
raw_nyc = pd.read_json("raw_nyc_phil.json")


In [19]:
raw_nyc.head()

programs
0  {'season': '1842-43', 'orchestra': 'New York P...
1  {'season': '1842-43', 'orchestra': 'New York P...
2  {'season': '1842-43', 'orchestra': 'Musicians ...
3  {'season': '1842-43', 'orchestra': 'New York P...
4  {'season': '1843-44', 'orchestra': 'New York P...

## Normalize a json data so that you get an output with columns

Date 	eventType 	Venue 	Location 	Time 	id 	orchestra 	programID 	season

In [30]:
pd.json_normalize(raw_nyc['programs'], record_path = 'concerts', meta = ['id','orchestra','programID','season'])


Date                      eventType  \
0      1842-12-07T05:00:00Z            Subscription Season   
1      1843-02-18T05:00:00Z            Subscription Season   
2      1843-04-07T05:00:00Z                        Special   
3      1843-04-22T05:00:00Z            Subscription Season   
4      1843-11-18T05:00:00Z            Subscription Season   
...                     ...                            ...   
21602  2017-06-24T04:00:00Z                        Chamber   
21603  2017-07-02T04:00:00Z                           Tour   
21604  2017-07-03T04:00:00Z                           Tour   
21605  2017-07-04T04:00:00Z  Tour - Young People's Concert   
21606  2017-07-07T04:00:00Z                           Tour   

                                      Venue          Location    Time  \
0                              Apollo Rooms     Manhattan, NY  8:00PM   
1                              Apollo Rooms     Manhattan, NY  8:00PM   
2                              Apollo Rooms     Manhattan, NY  8:00PM   
3                              Apollo Rooms     Manhattan, NY  8:00PM   
4                              Apollo Rooms     Manhattan, NY    None   
...                                     ...               ...     ...   
21602                            Guild Hall  East Hampton, NY    None   
21603                Shanghai Symphony Hall   Shanghai, CHINA  7:30PM   
21604                Shanghai Symphony Hall   Shanghai, CHINA  7:30PM   
21605  Shanghai Symphony Hall--Chamber Hall   Shanghai, CHINA  7:30PM   
21606                Shanghai Symphony Hall   Shanghai, CHINA  7:30PM   

                                         id  \
0      38e072a7-8fc9-4f9a-8eac-3957905c0002   
1      c7b2b95c-5e0b-431c-a340-5b37fc860b34   
2      894e1a52-1ae5-4fa7-aec0-b99997555a37   
3      34ec2c2b-3297-4716-9831-b538310462b7   
4      610a4acc-94e4-4cd6-bdc1-8ad020edc7e9   
...                                     ...   
21602  3d50968c-2e8b-405f-969c-ba36f941f393   
21603  adb9fc8b-95bc-47d0-9f4b-78fc242f1f8f   
21604  f072c652-c452-4ad4-975a-f6abd5ff8b4d   
21605  68625626-8085-44e4-bdb3-d8e6ac51c95b   
21606  5c22d35f-a282-4505-bd31-df619c74f52b   

                                      orchestra programID   season  
0                         New York Philharmonic      3853  1842-43  
1                         New York Philharmonic      5178  1842-43  
2      Musicians from the New York Philharmonic     10785  1842-43  
3                         New York Philharmonic      5887  1842-43  
4                         New York Philharmonic       305  1843-44  
...                                         ...       ...      ...  
21602  Musicians from the New York Philharmonic     14189  2016-17  
21603                     New York Philharmonic     14146  2016-17  
21604                     New York Philharmonic     14148  2016-17  
21605                     New York Philharmonic     14136  2016-17  
21606                     New York Philharmonic     14147  2016-17  

[21607 rows x 9 columns]

## Normalize a json data to get an output with columns :

soloistName 	soloistRoles 	soloistInstrument 	id

In [31]:
pd.json_normalize(raw_nyc["programs"], record_path = ['works', 'soloists'])


soloistName soloistRoles  \
0                                       Otto, Antoinette            S   
1                                  Scharfenberg, William            A   
2                                    Hill, Ureli Corelli            A   
3                                         Derwort, G. H.            A   
4                                        Boucher, Alfred            A   
...                                                  ...          ...   
56189                                         Xu, Weiqin            S   
56190                                Martin, Christopher            S   
56191                                     Shryock, Grace            S   
56192  Jazz at Lincoln Center Orchestra with Wynton M...            S   
56193  Jazz at Lincoln Center Orchestra with Wynton M...            S   

      soloistInstrument  
0               Soprano  
1                 Piano  
2                Violin  
3                 Viola  
4                 Cello  
...                 ...  
56189          Baritone  
56190           Trumpet  
56191      English Horn  
56192     Jazz Ensemble  
56193     Jazz Ensemble  

[56194 rows x 3 columns]

## Normalize a json to get an output with columns
soloistName 	soloistRoles 	soloistInstrument 	id 	orchestra 	programID 	season 	 	Date 	eventType 	Venue 	Location 	Time

In [34]:
a = pd.json_normalize(raw_nyc["programs"], record_path = ['works', 'soloists'], meta = ["id", "orchestra", "programID","season", "concerts"])

In [35]:
a.head()

soloistName soloistRoles soloistInstrument  \
0       Otto, Antoinette            S           Soprano   
1  Scharfenberg, William            A             Piano   
2    Hill, Ureli Corelli            A            Violin   
3         Derwort, G. H.            A             Viola   
4        Boucher, Alfred            A             Cello   

                                     id              orchestra programID  \
0  38e072a7-8fc9-4f9a-8eac-3957905c0002  New York Philharmonic      3853   
1  38e072a7-8fc9-4f9a-8eac-3957905c0002  New York Philharmonic      3853   
2  38e072a7-8fc9-4f9a-8eac-3957905c0002  New York Philharmonic      3853   
3  38e072a7-8fc9-4f9a-8eac-3957905c0002  New York Philharmonic      3853   
4  38e072a7-8fc9-4f9a-8eac-3957905c0002  New York Philharmonic      3853   

    season                                           concerts  
0  1842-43  [{'Date': '1842-12-07T05:00:00Z', 'eventType':...  
1  1842-43  [{'Date': '1842-12-07T05:00:00Z', 'eventType':...  
2  1842-43  [{'Date': '1842-12-07T05:00:00Z', 'eventType':...  
3  1842-43  [{'Date': '1842-12-07T05:00:00Z', 'eventType':...  
4  1842-43  [{'Date': '1842-12-07T05:00:00Z', 'eventType':...

In [47]:
concert = a['concerts'].apply(pd.Series)

In [49]:
dfconcert =pd.json_normalize(concert[0])
final = pd.concat([a, dfconcert], axis=1)

In [50]:
final.drop('concerts', axis= 1)

soloistName soloistRoles  \
0                                       Otto, Antoinette            S   
1                                  Scharfenberg, William            A   
2                                    Hill, Ureli Corelli            A   
3                                         Derwort, G. H.            A   
4                                        Boucher, Alfred            A   
...                                                  ...          ...   
56189                                         Xu, Weiqin            S   
56190                                Martin, Christopher            S   
56191                                     Shryock, Grace            S   
56192  Jazz at Lincoln Center Orchestra with Wynton M...            S   
56193  Jazz at Lincoln Center Orchestra with Wynton M...            S   

      soloistInstrument                                    id  \
0               Soprano  38e072a7-8fc9-4f9a-8eac-3957905c0002   
1                 Piano  38e072a7-8fc9-4f9a-8eac-3957905c0002   
2                Violin  38e072a7-8fc9-4f9a-8eac-3957905c0002   
3                 Viola  38e072a7-8fc9-4f9a-8eac-3957905c0002   
4                 Cello  38e072a7-8fc9-4f9a-8eac-3957905c0002   
...                 ...                                   ...   
56189          Baritone  68625626-8085-44e4-bdb3-d8e6ac51c95b   
56190           Trumpet  5c22d35f-a282-4505-bd31-df619c74f52b   
56191      English Horn  5c22d35f-a282-4505-bd31-df619c74f52b   
56192     Jazz Ensemble  5c22d35f-a282-4505-bd31-df619c74f52b   
56193     Jazz Ensemble  5c22d35f-a282-4505-bd31-df619c74f52b   

                   orchestra programID   season                  Date  \
0      New York Philharmonic      3853  1842-43  1842-12-07T05:00:00Z   
1      New York Philharmonic      3853  1842-43  1842-12-07T05:00:00Z   
2      New York Philharmonic      3853  1842-43  1842-12-07T05:00:00Z   
3      New York Philharmonic      3853  1842-43  1842-12-07T05:00:00Z   
4      New York Philharmonic      3853  1842-43  1842-12-07T05:00:00Z   
...                      ...       ...      ...                   ...   
56189  New York Philharmonic     14136  2016-17  2017-07-04T04:00:00Z   
56190  New York Philharmonic     14147  2016-17  2017-07-07T04:00:00Z   
56191  New York Philharmonic     14147  2016-17  2017-07-07T04:00:00Z   
56192  New York Philharmonic     14147  2016-17  2017-07-07T04:00:00Z   
56193  New York Philharmonic     14147  2016-17  2017-07-07T04:00:00Z   

                           eventType                                 Venue  \
0                Subscription Season                          Apollo Rooms   
1                Subscription Season                          Apollo Rooms   
2                Subscription Season                          Apollo Rooms   
3                Subscription Season                          Apollo Rooms   
4                Subscription Season                          Apollo Rooms   
...                              ...                                   ...   
56189  Tour - Young People's Concert  Shanghai Symphony Hall--Chamber Hall   
56190                           Tour                Shanghai Symphony Hall   
56191                           Tour                Shanghai Symphony Hall   
56192                           Tour                Shanghai Symphony Hall   
56193                           Tour                Shanghai Symphony Hall   

              Location    Time  
0        Manhattan, NY  8:00PM  
1        Manhattan, NY  8:00PM  
2        Manhattan, NY  8:00PM  
3        Manhattan, NY  8:00PM  
4        Manhattan, NY  8:00PM  
...                ...     ...  
56189  Shanghai, CHINA  7:30PM  
56190  Shanghai, CHINA  7:30PM  
56191  Shanghai, CHINA  7:30PM  
56192  Shanghai, CHINA  7:30PM  
56193  Shanghai, CHINA  7:30PM  

[56194 rows x 12 columns]

# euro_cup and read the data
-Euro_2012_stats_TEAM.csv

In [2]:
euro_cup = pd.read_csv("Euro_2012_stats_TEAM.csv")
euro_cup.head()

Team  Goals  Shots on target  Shots off target Shooting Accuracy  \
0         Croatia      4               13                12             51.9%   
1  Czech Republic      4               13                18             41.9%   
2         Denmark      4               10                10             50.0%   
3         England      5               11                18             50.0%   
4          France      3               22                24             37.9%   

  % Goals-to-shots  Total shots (inc. Blocked)  Hit Woodwork  Penalty goals  \
0            16.0%                          32             0              0   
1            12.9%                          39             0              0   
2            20.0%                          27             1              0   
3            17.2%                          40             0              0   
4             6.5%                          65             1              0   

   Penalties not scored  ...  Saves made  Saves-to-shots ratio  Fouls Won  \
0                     0  ...          13                 81.3%         41   
1                     0  ...           9                 60.1%         53   
2                     0  ...          10                 66.7%         25   
3                     0  ...          22                 88.1%         43   
4                     0  ...           6                 54.6%         36   

  Fouls Conceded  Offsides  Yellow Cards  Red Cards  Subs on  Subs off  \
0             62         2             9          0        9         9   
1             73         8             7          0       11        11   
2             38         8             4          0        7         7   
3             45         6             5          0       11        11   
4             51         5             6          0       11        11   

   Players Used  
0            16  
1            19  
2            15  
3            16  
4            19  

[5 rows x 35 columns]

## select the goal Column

In [3]:
euro_cup["Goals"]

0      4
1      4
2      4
3      5
4      3
5     10
6      5
7      6
8      2
9      2
10     6
11     1
12     5
13    12
14     5
15     2
Name: Goals, dtype: int64

## How many team participated in the Euro2012?

In [4]:
euro_cup["Team"].nunique()

16

##  What is the number of columns in the dataset?

In [5]:
euro_cup.columns.nunique()

35

## View only the columns Team, Yellow Cards and Red Cards and assign them to a dataframe called discipline

In [6]:
discipline = euro_cup[["Team", "Yellow Cards", "Red Cards"]]
discipline

Team  Yellow Cards  Red Cards
0               Croatia             9          0
1        Czech Republic             7          0
2               Denmark             4          0
3               England             5          0
4                France             6          0
5               Germany             4          0
6                Greece             9          1
7                 Italy            16          0
8           Netherlands             5          0
9                Poland             7          1
10             Portugal            12          0
11  Republic of Ireland             6          1
12               Russia             6          0
13                Spain            11          0
14               Sweden             7          0
15              Ukraine             5          0

## Sort the teams by Red Cards, then to Yellow Cards

In [7]:
discipline.sort_values(by = ["Red Cards", "Yellow Cards"])

Team  Yellow Cards  Red Cards
2               Denmark             4          0
5               Germany             4          0
3               England             5          0
8           Netherlands             5          0
15              Ukraine             5          0
4                France             6          0
12               Russia             6          0
1        Czech Republic             7          0
14               Sweden             7          0
0               Croatia             9          0
13                Spain            11          0
10             Portugal            12          0
7                 Italy            16          0
11  Republic of Ireland             6          1
9                Poland             7          1
6                Greece             9          1

## Calculate the mean Yellow Cards given per Team

In [9]:
discipline["Yellow Cards"].mean()

7.4375

##  Filter teams that scored more than 6 goals

In [10]:
euro_cup[euro_cup["Goals"] > 6][["Team", "Goals"]]

Team  Goals
5   Germany     10
13    Spain     12

## Select the teams that start with G

In [11]:
euro_cup[euro_cup["Team"].str.startswith("G")]

Team  Goals  Shots on target  Shots off target Shooting Accuracy  \
5  Germany     10               32                32             47.8%   
6   Greece      5                8                18             30.7%   

  % Goals-to-shots  Total shots (inc. Blocked)  Hit Woodwork  Penalty goals  \
5            15.6%                          80             2              1   
6            19.2%                          32             1              1   

   Penalties not scored  ...  Saves made  Saves-to-shots ratio  Fouls Won  \
5                     0  ...          10                 62.6%         63   
6                     1  ...          13                 65.1%         67   

  Fouls Conceded  Offsides  Yellow Cards  Red Cards  Subs on  Subs off  \
5             49        12             4          0       15        15   
6             48        12             9          1       12        12   

   Players Used  
5            17  
6            20  

[2 rows x 35 columns]

##  Select the first 7 columns

In [12]:
euro_cup.iloc[:, :7]

Team  Goals  Shots on target  Shots off target  \
0               Croatia      4               13                12   
1        Czech Republic      4               13                18   
2               Denmark      4               10                10   
3               England      5               11                18   
4                France      3               22                24   
5               Germany     10               32                32   
6                Greece      5                8                18   
7                 Italy      6               34                45   
8           Netherlands      2               12                36   
9                Poland      2               15                23   
10             Portugal      6               22                42   
11  Republic of Ireland      1                7                12   
12               Russia      5                9                31   
13                Spain     12               42                33   
14               Sweden      5               17                19   
15              Ukraine      2                7                26   

   Shooting Accuracy % Goals-to-shots  Total shots (inc. Blocked)  
0              51.9%            16.0%                          32  
1              41.9%            12.9%                          39  
2              50.0%            20.0%                          27  
3              50.0%            17.2%                          40  
4              37.9%             6.5%                          65  
5              47.8%            15.6%                          80  
6              30.7%            19.2%                          32  
7              43.0%             7.5%                         110  
8              25.0%             4.1%                          60  
9              39.4%             5.2%                          48  
10             34.3%             9.3%                          82  
11             36.8%             5.2%                          28  
12             22.5%            12.5%                          59  
13             55.9%            16.0%                         100  
14             47.2%            13.8%                          39  
15             21.2%             6.0%                          38

## Select all columns except the last 3.

In [13]:
euro_cup.iloc[:, :-3]

Team  Goals  Shots on target  Shots off target  \
0               Croatia      4               13                12   
1        Czech Republic      4               13                18   
2               Denmark      4               10                10   
3               England      5               11                18   
4                France      3               22                24   
5               Germany     10               32                32   
6                Greece      5                8                18   
7                 Italy      6               34                45   
8           Netherlands      2               12                36   
9                Poland      2               15                23   
10             Portugal      6               22                42   
11  Republic of Ireland      1                7                12   
12               Russia      5                9                31   
13                Spain     12               42                33   
14               Sweden      5               17                19   
15              Ukraine      2                7                26   

   Shooting Accuracy % Goals-to-shots  Total shots (inc. Blocked)  \
0              51.9%            16.0%                          32   
1              41.9%            12.9%                          39   
2              50.0%            20.0%                          27   
3              50.0%            17.2%                          40   
4              37.9%             6.5%                          65   
5              47.8%            15.6%                          80   
6              30.7%            19.2%                          32   
7              43.0%             7.5%                         110   
8              25.0%             4.1%                          60   
9              39.4%             5.2%                          48   
10             34.3%             9.3%                          82   
11             36.8%             5.2%                          28   
12             22.5%            12.5%                          59   
13             55.9%            16.0%                         100   
14             47.2%            13.8%                          39   
15             21.2%             6.0%                          38   

    Hit Woodwork  Penalty goals  Penalties not scored  ...  Clean Sheets  \
0              0              0                     0  ...             0   
1              0              0                     0  ...             1   
2              1              0                     0  ...             1   
3              0              0                     0  ...             2   
4              1              0                     0  ...             1   
5              2              1                     0  ...             1   
6              1              1                     1  ...             1   
7              2              0                     0  ...             2   
8              2              0                     0  ...             0   
9              0              0                     0  ...             0   
10             6              0                     0  ...             2   
11             0              0                     0  ...             0   
12             2              0                     0  ...             0   
13             0              1                     0  ...             5   
14             3              0                     0  ...             1   
15             0              0                     0  ...             0   

    Blocks  Goals conceded Saves made  Saves-to-shots ratio  Fouls Won  \
0       10               3         13                 81.3%         41   
1       10               6          9                 60.1%         53   
2       10               5         10                 66.7%         25   
3       29               3         22                 88.1%         43   
4       

## Present only the Shooting Accuracy from England, Italy and Russia

In [14]:
euro_cup[euro_cup["Team"].isin(["England", "Italy", "Russia"])][["Team", "Shooting Accuracy"]]

Team Shooting Accuracy
3   England             50.0%
7     Italy             43.0%
12   Russia             22.5%

## Sort by the name of the country in descending

In [16]:
discipline.sort_values(by= "Team", ascending = False)["Team"]

15                Ukraine
14                 Sweden
13                  Spain
12                 Russia
11    Republic of Ireland
10               Portugal
9                  Poland
8             Netherlands
7                   Italy
6                  Greece
5                 Germany
4                  France
3                 England
2                 Denmark
1          Czech Republic
0                 Croatia
Name: Team, dtype: object

## Step 7. What was the country that scored highest goal?

In [17]:
euro_cup[euro_cup["Goals"]==max(euro_cup["Goals"])][["Team", "Goals"]]

Team  Goals
13  Spain     12

In [18]:
euro_cup[(euro_cup.Goals == max(euro_cup.Goals))].Team


13    Spain
Name: Team, dtype: object

In [19]:
euro_cup.query("Goals == Goals.max()")["Team"]

13    Spain
Name: Team, dtype: object

## remove all the percentage sign from the shotting accuracy column that has a percentage sign

In [20]:
euro_cup["Shooting Accuracy"]= euro_cup["Shooting Accuracy"].str.replace("%", "")
euro_cup.head(3)

Team  Goals  Shots on target  Shots off target Shooting Accuracy  \
0         Croatia      4               13                12              51.9   
1  Czech Republic      4               13                18              41.9   
2         Denmark      4               10                10              50.0   

  % Goals-to-shots  Total shots (inc. Blocked)  Hit Woodwork  Penalty goals  \
0            16.0%                          32             0              0   
1            12.9%                          39             0              0   
2            20.0%                          27             1              0   

   Penalties not scored  ...  Saves made  Saves-to-shots ratio  Fouls Won  \
0                     0  ...          13                 81.3%         41   
1                     0  ...           9                 60.1%         53   
2                     0  ...          10                 66.7%         25   

  Fouls Conceded  Offsides  Yellow Cards  Red Cards  Subs on  Subs off  \
0             62         2             9          0        9         9   
1             73         8             7          0       11        11   
2             38         8             4          0        7         7   

   Players Used  
0            16  
1            19  
2            15  

[3 rows x 35 columns]

# Read wine data

In [25]:
wine = pd.read_csv("wine.data", header = None)
wine

0      1     2     3     4    5     6     7     8     9      10    11  \
0     1  14.23  1.71  2.43  15.6  127  2.80  3.06  0.28  2.29   5.64  1.04   
1     1  13.20  1.78  2.14  11.2  100  2.65  2.76  0.26  1.28   4.38  1.05   
2     1  13.16  2.36  2.67  18.6  101  2.80  3.24  0.30  2.81   5.68  1.03   
3     1  14.37  1.95  2.50  16.8  113  3.85  3.49  0.24  2.18   7.80  0.86   
4     1  13.24  2.59  2.87  21.0  118  2.80  2.69  0.39  1.82   4.32  1.04   
..   ..    ...   ...   ...   ...  ...   ...   ...   ...   ...    ...   ...   
173   3  13.71  5.65  2.45  20.5   95  1.68  0.61  0.52  1.06   7.70  0.64   
174   3  13.40  3.91  2.48  23.0  102  1.80  0.75  0.43  1.41   7.30  0.70   
175   3  13.27  4.28  2.26  20.0  120  1.59  0.69  0.43  1.35  10.20  0.59   
176   3  13.17  2.59  2.37  20.0  120  1.65  0.68  0.53  1.46   9.30  0.60   
177   3  14.13  4.10  2.74  24.5   96  2.05  0.76  0.56  1.35   9.20  0.61   

       12    13  
0    3.92  1065  
1    3.40  1050  
2    3.17  1185  
3    3.45  1480  
4    2.93   735  
..    ...   ...  
173  1.74   740  
174  1.56   750  
175  1.56   835  
176  1.62   840  
177  1.60   560  

[178 rows x 14 columns]

## Delete the first, fourth, seventh, nineth, eleventh, thirteenth and fourteenth columns

In [26]:
wine.drop(wine.columns[[0,3,6,8,10,12,13]], axis = 1, inplace = True)

In [27]:
wine.head()

1     2     4    5     7     9     11
0  14.23  1.71  15.6  127  3.06  2.29  1.04
1  13.20  1.78  11.2  100  2.76  1.28  1.05
2  13.16  2.36  18.6  101  3.24  2.81  1.03
3  14.37  1.95  16.8  113  3.49  2.18  0.86
4  13.24  2.59  21.0  118  2.69  1.82  1.04

## Step 5. Assign the columns as below:


1) alcohol  
2) malic_acid  
3) alcalinity_of_ash  
4) magnesium  
5) flavanoids  
6) proanthocyanins  
7) hue 

In [28]:
wine.rename(columns = {1:"alcohol", 2:"malic_acid", 4:"alcalinity_of_ash", 5: "magnesium",
                      7:"flavanoids", 9:"proanthocyanins", 11: "hue"}, inplace = True)

In [29]:
wine.columns = ['alcohol','malic_acid', 'alcalinity_of_ash', 'magnesium','flavanoids','proanthocyanins','hue']
wine.head()

alcohol  malic_acid  alcalinity_of_ash  magnesium  flavanoids  \
0    14.23        1.71               15.6        127        3.06   
1    13.20        1.78               11.2        100        2.76   
2    13.16        2.36               18.6        101        3.24   
3    14.37        1.95               16.8        113        3.49   
4    13.24        2.59               21.0        118        2.69   

   proanthocyanins   hue  
0             2.29  1.04  
1             1.28  1.05  
2             2.81  1.03  
3             2.18  0.86  
4             1.82  1.04

## Set the values of the first 3 rows from alcohol as NaN


In [30]:
wine.loc[:2, "alcohol"] = "NaN"

In [31]:
wine.head()

alcohol  malic_acid  alcalinity_of_ash  magnesium  flavanoids  \
0     NaN        1.71               15.6        127        3.06   
1     NaN        1.78               11.2        100        2.76   
2     NaN        2.36               18.6        101        3.24   
3   14.37        1.95               16.8        113        3.49   
4   13.24        2.59               21.0        118        2.69   

   proanthocyanins   hue  
0             2.29  1.04  
1             1.28  1.05  
2             2.81  1.03  
3             2.18  0.86  
4             1.82  1.04

## Now set the value of the rows 3 and 4 of magnesium as NaN¶

In [32]:
wine.loc[2:3, "magnesium"] = "NaN"

In [33]:
wine.head()

alcohol  malic_acid  alcalinity_of_ash magnesium  flavanoids  \
0     NaN        1.71               15.6       127        3.06   
1     NaN        1.78               11.2       100        2.76   
2     NaN        2.36               18.6       NaN        3.24   
3   14.37        1.95               16.8       NaN        3.49   
4   13.24        2.59               21.0       118        2.69   

   proanthocyanins   hue  
0             2.29  1.04  
1             1.28  1.05  
2             2.81  1.03  
3             2.18  0.86  
4             1.82  1.04

## Fill the value of NaN with the number 10 in alcohol and 100 in magnesium

In [34]:
wine.loc[:2, "alcohol"] = 10
wine.loc[2:3, "magnesium"] = 100
wine.head()

alcohol  malic_acid  alcalinity_of_ash magnesium  flavanoids  \
0      10        1.71               15.6       127        3.06   
1      10        1.78               11.2       100        2.76   
2      10        2.36               18.6       100        3.24   
3   14.37        1.95               16.8       100        3.49   
4   13.24        2.59               21.0       118        2.69   

   proanthocyanins   hue  
0             2.29  1.04  
1             1.28  1.05  
2             2.81  1.03  
3             2.18  0.86  
4             1.82  1.04

In [35]:
wine.fillna({"alcohol":10,"magnesium":100},inplace=True)
wine

alcohol  malic_acid  alcalinity_of_ash  magnesium  flavanoids  \
0      10.00        1.71               15.6        127        3.06   
1      10.00        1.78               11.2        100        2.76   
2      10.00        2.36               18.6        100        3.24   
3      14.37        1.95               16.8        100        3.49   
4      13.24        2.59               21.0        118        2.69   
..       ...         ...                ...        ...         ...   
173    13.71        5.65               20.5         95        0.61   
174    13.40        3.91               23.0        102        0.75   
175    13.27        4.28               20.0        120        0.69   
176    13.17        2.59               20.0        120        0.68   
177    14.13        4.10               24.5         96        0.76   

     proanthocyanins   hue  
0               2.29  1.04  
1               1.28  1.05  
2               2.81  1.03  
3               2.18  0.86  
4               1.82  1.04  
..               ...   ...  
173             1.06  0.64  
174             1.41  0.70  
175             1.35  0.59  
176             1.46  0.60  
177             1.35  0.61  

[178 rows x 7 columns]

## Count the number of missing values

In [36]:
wine.isnull().sum()

alcohol              0
malic_acid           0
alcalinity_of_ash    0
magnesium            0
flavanoids           0
proanthocyanins      0
hue                  0
dtype: int64

## Create an array of 10 random numbers up until 10

In [37]:
arr = np.random.randint(low = 0, high = 10, size = (10))
print(arr)

[5 3 3 3 2 0 0 4 8 2]


## random numbers you generated as an index and assign NaN value to each of cell.

In [38]:
#  made a duplicate of the wine data frame
wine1 = wine.copy()
wine1.head()

alcohol  malic_acid  alcalinity_of_ash  magnesium  flavanoids  \
0    10.00        1.71               15.6        127        3.06   
1    10.00        1.78               11.2        100        2.76   
2    10.00        2.36               18.6        100        3.24   
3    14.37        1.95               16.8        100        3.49   
4    13.24        2.59               21.0        118        2.69   

   proanthocyanins   hue  
0             2.29  1.04  
1             1.28  1.05  
2             2.81  1.03  
3             2.18  0.86  
4             1.82  1.04

In [39]:
# assigning the NaN value based on the random index value
wine1.loc[arr] = None
wine1.head()

alcohol  malic_acid  alcalinity_of_ash  magnesium  flavanoids  \
0      NaN         NaN                NaN        NaN         NaN   
1     10.0        1.78               11.2      100.0        2.76   
2      NaN         NaN                NaN        NaN         NaN   
3      NaN         NaN                NaN        NaN         NaN   
4      NaN         NaN                NaN        NaN         NaN   

   proanthocyanins   hue  
0              NaN   NaN  
1             1.28  1.05  
2              NaN   NaN  
3              NaN   NaN  
4              NaN   NaN

## How many missing values do we have?

In [40]:
wine1.isna().sum()

alcohol              6
malic_acid           6
alcalinity_of_ash    6
magnesium            6
flavanoids           6
proanthocyanins      6
hue                  6
dtype: int64

## Delete the rows that contain missing values

In [41]:
wine2 = wine1.dropna()

In [42]:
wine2.isnull().sum()

alcohol              0
malic_acid           0
alcalinity_of_ash    0
magnesium            0
flavanoids           0
proanthocyanins      0
hue                  0
dtype: int64

## Print only the non-null values in alcohol

In [43]:
wine2[wine2["alcohol"].notnull()]["alcohol"]

1      10.00
6      14.39
7      14.06
9      13.86
10     14.10
       ...  
173    13.71
174    13.40
175    13.27
176    13.17
177    14.13
Name: alcohol, Length: 172, dtype: float64

## Reset the index, so it starts with 0 again

In [44]:
wine2.reset_index()

index  alcohol  malic_acid  alcalinity_of_ash  magnesium  flavanoids  \
0        1    10.00        1.78               11.2      100.0        2.76   
1        6    14.39        1.87               14.6       96.0        2.52   
2        7    14.06        2.15               17.6      121.0        2.51   
3        9    13.86        1.35               16.0       98.0        3.15   
4       10    14.10        2.16               18.0      105.0        3.32   
..     ...      ...         ...                ...        ...         ...   
167    173    13.71        5.65               20.5       95.0        0.61   
168    174    13.40        3.91               23.0      102.0        0.75   
169    175    13.27        4.28               20.0      120.0        0.69   
170    176    13.17        2.59               20.0      120.0        0.68   
171    177    14.13        4.10               24.5       96.0        0.76   

     proanthocyanins   hue  
0               1.28  1.05  
1               1.98  1.02  
2               1.25  1.06  
3               1.85  1.01  
4               2.38  1.25  
..               ...   ...  
167             1.06  0.64  
168             1.41  0.70  
169             1.35  0.59  
170             1.46  0.60  
171             1.35  0.61  

[172 rows x 8 columns]